apm-agent-attach-cli/src/main/java/co/elastic/apm/attach/AgentDownloader.java

```diff
@@ -0,0 +1,190 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.attach;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Locale;
+
+import static co.elastic.apm.attach.Utils.getTargetAgentDir;
+
+/**
+ * A utility for downloading any given version of the Elastic APM Java agent from maven central.
+ * After being downloaded, the agent jar is verified to match the expected PGP signature from maven central. If
+ * verification fails, the agent is removed from the local storage and the entire task fails.
+ */
+public class AgentDownloader {
+
+    private static final String AGENT_GROUP_ID = "co.elastic.apm";
+    private static final String AGENT_ARTIFACT_ID = "elastic-apm-agent";
+
+    // intentionally not static so that we create the logger only after proper initialization
+    private final Logger logger = LogManager.getLogger(AgentDownloader.class);
+
+    public AgentDownloader(PgpSignatureVerifier pgpSignatureVerifier) {
+        this.pgpSignatureVerifier = pgpSignatureVerifier;
+    }
+
+    private final PgpSignatureVerifier pgpSignatureVerifier;
+
+    /**
+     * Downloads the agent jar, authenticates and verifies its PGP signature and returns the path for the locally
+     * stored jar.
+     * @param agentVersion the target agent version
+     * @return the path for the locally stored agent jar
+     * @throws Exception failure either with downloading, copying to local file system, or in PGP signature verification
+     */
+    Path downloadAndVerifyAgent(String agentVersion) throws Exception {
+        logger.debug("Requested to download Elastic APM Java agent version {}", agentVersion);
+        final String mavenAgentBaseUrl = getAgentMavenBaseUrl(agentVersion);
+        Path targetDir = getTargetAgentDir(agentVersion);
+        String agentJarName = computeAgentJarName(agentVersion);
+        String mavenAgentJarUrl = computeFileUrl(mavenAgentBaseUrl, agentJarName);
+        Path localAgentJarPath = targetDir.resolve(agentJarName);
+        boolean downloadAndVerify = true;
+
+        if (Files.exists(localAgentJarPath)) {
+            logger.debug("{} already exists at {}", agentJarName, targetDir);
+            if (Files.isReadable(localAgentJarPath)) {
+                downloadAndVerify = false;
+            } else {
+                // quick and dirty - this is a temporary solution until we download agent from the Fleet package registry
+                logger.info("{} is not readable for the current user, removing and downloading again", localAgentJarPath);
+                Files.delete(localAgentJarPath);
+            }
+        }
+
+        if (downloadAndVerify) {
```

Comment:
If file exists, we should still check signature, otherwise that's a way to inject untrusted code, skipping download might be still relevant though. Also, if there is any download error and file is corrupted, deleting the file would be a better option otherwise that would require manual cleanup.

Reply:
> If file exists, we should still check signature, otherwise that's a way to inject untrusted code

Not really, if attackers have access to the server host, they can do whatever they want, including replacing the cli jar with malicious one. Similarly, we cannot control the verification of the cli jar itself, so if the user doesn't make sure to download it from a trusted place, it may contain malicious contents.

As far as I see it, we are only dealing with man-in-the-middle attacks here - make sure our cli tool downloads only the trusted agent jar.

> Also, if there is any download error and file is corrupted, deleting the file would be a better option otherwise that would require manual cleanup.

You mean - download has an error but ends without exception? Is that a case we need to consider?

Reply:
summary of what has been discussed offline:
- when downloading, we should always check the signature, to prevent against "download successful but with wrong content"
- when file is already present, checking the signature of file would only detect if local file has been tampered with when we try to download it, but:
  a) we have no guarantee that attacher with download option will be called
  b) anyone having access to the folder where the attacher is could have replaced both the attacher and agent
  c) it is probably a better responsibility for the Fleet Agent to check for binary integrity (both Java agent and the cli-attacher, which could be implicit for the latter if embedded in Fleet agent).